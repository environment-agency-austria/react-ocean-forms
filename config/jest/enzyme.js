import Enzyme from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';
import 'babel-polyfill';

// Currently there's no official enzyme support
// for reacts context api. FIXME
// https://github.com/airbnb/enzyme/issues/1553
import Adapter from './enzymeAdapter';

Enzyme.configure({ adapter: new Adapter() });
