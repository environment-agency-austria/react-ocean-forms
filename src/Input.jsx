/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';

import FieldLine from './FieldLine';
import { fieldMetaShape, fieldShape } from './shapes';

/**
 * Component for displaying bootstrap
 * form groups with an html input and
 * oForm support
 */
function Input(props) {
  const {
    field,
    type,
    meta,
  } = props;

  return (
    <FieldLine {...props}>
      {meta.plaintext ? field.value : <input type={type} {...field} />}
    </FieldLine>
  );
}

Input.displayName = 'Input';

Input.defaultProps = {
  type: 'text',
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
};

export default Input;
