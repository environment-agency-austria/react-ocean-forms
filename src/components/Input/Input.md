### Examples
#### Input types
Showcase of the input type support

```jsx
import { Form, Input } from 'react-ocean-forms';

function Example() {
  // onChange callback
  const handleChange = (value) => {
    console.log('onChange, value: ' + JSON.stringify(value));
  };

  // onBlur callback
  const handleBlur = () => {
    console.log('onBlur');
  };

  return (
    <Form className="demo">
      <Input
        name="myInput"
        label="Number input"
        onChange={handleChange}
        onBlur={handleBlur}
        type="number"
      />

      <Input
        name="myInput2"
        label="Date input"
        onChange={handleChange}
        onBlur={handleBlur}
        type="date"
      />
    </Form>
  );
}

<Example />
```

#### Custom input
Creating your own input components is quite simple. Note that if you want to use the flexible Intl support, you should put all your text outputs through meta.stringFormatter.

```jsx
import { Form, withField } from 'react-ocean-forms';

/**
 * Simple implementation of a custom input
 */
function BaseCustomInput(props) {
  const {
    field,
    label,
    type,
    meta,
    customProp,
  } = props;


  return (
    <div>
      <div>
        <label htmlFor={field.id}>
          <strong>{meta.stringFormatter(label)}</strong>
        </label>
      </div>
      <input type={type} {...field} />
      {customProp}
    </div>
  );
}
const CustomInput = withField(BaseCustomInput);

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = (values) => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // onChange callback
  const handleChange = (value) => {
    console.log('onChange, value: ' + JSON.stringify(value));
  };

  // onBlur callback
  const handleBlur = () => {
    console.log('onBlur');
  };

  return (
    <Form className="demo" onSubmit={handleSubmit}>
      <CustomInput
        name="myCustomInput"
        label="Input with info text"
        onChange={handleChange}
        onBlur={handleBlur}
        customProp="custom text"
      />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />
```
