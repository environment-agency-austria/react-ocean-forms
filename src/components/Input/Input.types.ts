/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentProps } from '../Field';

/**
 * Props for the Input component
 */
export interface IInputProps extends IFieldComponentProps {
  /**
   * Input type (e.g. text, number, ...)
   */
  type: string;
}
