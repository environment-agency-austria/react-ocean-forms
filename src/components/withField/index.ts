export * from './withField';
export * from './Field/Field.types';
