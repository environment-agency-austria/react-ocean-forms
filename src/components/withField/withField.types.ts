/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta } from '../../hooks';

/**
 * Props for the input component of a Field
 */
export interface IFieldComponentProps {
  /**
   * Props for the actual html input, designed
   * to be passed as-is
   */
  field: IFieldComponentFieldProps;
  /**
   * Meta informations about the field state
   */
  meta: IFieldComponentMeta;
  /**
   * Label (string or message id)
   */
  label: string;
}
