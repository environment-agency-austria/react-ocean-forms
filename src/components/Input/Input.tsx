/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { useField } from '../../hooks';
import { FieldLine } from '../FieldLine';
import { IInputProps } from './Input.types';

/**
 * Defines a form line containing a label and an input. Additionally it
 * will render validation messages. If the user adds the required validator
 * then it will mark the field as required as well.
 */
export const Input: React.FC<IInputProps> = (props) => {
  const {
    type = 'text',
    ...rest
  } = props;

  const { fieldProps, metaProps } = useField(rest);

  const fieldValue = fieldProps.value;
  if (typeof fieldValue !== 'string' && fieldValue !== undefined) {
    throw new Error(
      'Incompatible field value supplied for input component '
      + `${fieldProps.id}. Only values with type string or undefined are allowed.`,
    );
  }

  return (
    <FieldLine {...props} field={fieldProps} meta={metaProps}>
      {metaProps.plaintext ? fieldProps.value : <input type={type} {...fieldProps} value={fieldValue} />}
    </FieldLine>
  );
}
