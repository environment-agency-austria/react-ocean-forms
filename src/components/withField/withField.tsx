/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { useField, IBaseFieldProps, IUseValidationArgs } from '../../hooks';
import { IFieldComponentProps } from './withField.types';

type WrappedValidatedComponentProps<TComp> =
  Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IFieldComponentProps> & IBaseFieldProps & IUseValidationArgs;

/**
 * Higher order component for validation
 * @deprecated Deprecated in favour of `useField` hook
 */
export const withField = <TComp extends React.ComponentType<TProps>, TProps extends IFieldComponentProps = PropsOf<TComp>>
(component: TComp): React.ComponentType<WrappedValidatedComponentProps<TComp>> => {

  const CastedComponent = component as React.ComponentType<TProps>;

  type IWrappedProps = WrappedValidatedComponentProps<TComp>;

  const ValidatedComponent: React.SFC<IWrappedProps> = (props: IWrappedProps): JSX.Element => {
    const { fieldProps, metaProps } = useField(props);

    return (
      // @ts-ignore
      <CastedComponent field={fieldProps} meta={metaProps} {...props} />
    );
  };
  ValidatedComponent.displayName = `withField(${getDisplayName(component)})`;

  return ValidatedComponent;
};
