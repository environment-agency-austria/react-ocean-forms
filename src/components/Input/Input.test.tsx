import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { IFieldComponentFieldProps, IFieldComponentMeta } from '../withField';
import { BaseInput } from './Input';
import { IInputProps } from './Input.types';

describe('<Input />', () => {
  interface ISetupArgs {
    props?: Partial<IInputProps>;
    metaOverrides?: Partial<IFieldComponentMeta>;
    fieldOverrides?: Partial<IFieldComponentFieldProps>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
  }

  const setup = ({
    props,
    metaOverrides,
    fieldOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const meta = {
      valid: true,
      error: null,
      isValidating: false,
      isRequired: false,
      touched: false,
      stringFormatter: jest.fn(),
      plaintext: false,
      ...metaOverrides,
    };
    const field = {
      id: 'unitInput',
      name: 'unitInput',
      value: '',
      disabled: false,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      ...fieldOverrides,
    };

    const wrapper = shallow((
      <BaseInput
        label="unitLabel"
        meta={meta}
        field={field}
        {...props}
      />
    ));

    return {
      wrapper,
    };
  };

  it('should render without crashing', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should only display the value if plaintext is set', () => {
    const wrapper = setup({ metaOverrides: { plaintext: true }});
    expect(wrapper).toMatchSnapshot();
  });

  describe('Unsupported value types', () => {
    it('should throw an error if the value is not a string', () => {
      expect(() => setup({ fieldOverrides: { value: 42 }})).toThrowError();
    });
  });
});
