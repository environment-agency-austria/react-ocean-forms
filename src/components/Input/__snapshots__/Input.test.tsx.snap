// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should only display the value if plaintext is set 1`] = `
Object {
  "wrapper": <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isRequired": false,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    name="unitInput"
  />,
}
`;

exports[`<Input /> should properly pass the type prop 1`] = `
Object {
  "wrapper": <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isRequired": false,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    name="unitInput"
    type="number"
  >
    <input
      disabled={false}
      id="unitInput"
      name="unitInput"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      type="number"
      value=""
    />
  </FieldLine>,
}
`;

exports[`<Input /> should render without crashing 1`] = `
Object {
  "wrapper": <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isRequired": false,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    name="unitInput"
  >
    <input
      disabled={false}
      id="unitInput"
      name="unitInput"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      type="text"
      value=""
    />
  </FieldLine>,
}
`;
